README.md
#Famous-Slideshow-Webpack
> A demo slideshow with Webpack and Famo.us. 

###Features

* This is the original slideshow from Famo.us wrapped in Webpack
* Built using the [famous-webpack-seed](https://github.com/talves/famous-webpack-seed)

---

### Getting started with Famo.us and webpack

####Installation

```bash
npm install -g webpack webpack-dev-server # install webpack
git clone https://github.com/talves/famous-slideshow-webpack # clone this repository
# rm -rf .git # optionally remove git history
npm install # install dependencies
```

####Development

```bash
webpack-dev-server --reload=localhost
```

Now navigate to:

* [http://localhost:8080/webpack-dev-server/index.html](http://localhost:8080/webpack-dev-server/index.html) (uses index.html)
* [http://localhost:8080/webpack-dev-server](http://localhost:8080/webpack-dev-server) (list the bundle)


####Production
```bash
webpack --minify --env=production
```
puts the build into the dist folder

* The optional `--minify` flag minifies the output.
* The optional `--env=xxx` flag sets a global `ENV` variable (default: `window.TARGET='dev'`).

## Contributors

Like it? Show some love and star this project!


CHANGELOG.md
# Changelog

### 0.1.0 (09/17/2014)

* fixed the loading of the context size
* uses the picasa search for images of puppies


src/views/AppView.js
define(function(require, exports, module) {
    var View = require('famous/core/View');
    var Surface = require('famous/core/Surface');
    var Transform = require('famous/core/Transform');
    var StateModifier = require('famous/modifiers/StateModifier');
    var ImageSurface = require('famous/surfaces/ImageSurface');
    var ContainerSurface = require('famous/surfaces/ContainerSurface');

    var SlideshowView = require('./SlideshowView');

    function AppView() {
        View.apply(this, arguments);

        this.options.cameraWidth = this.options.cameraWidth || 400;

        this.options.slideWidth = 0.8 * this.options.cameraWidth;
        this.options.slideHeight = this.options.slideWidth + 40;
        this.options.slidePosition = 0.77 * this.options.cameraWidth;

        _createCamera.call(this);

    }

    AppView.prototype = Object.create(View.prototype);
    AppView.prototype.constructor = AppView;

    AppView.DEFAULT_OPTIONS = {
        data: undefined,
        cameraWidth: undefined
    };

    function _createCamera() {
        var camera = new ImageSurface({
            size: [this.options.cameraWidth, true],
            content: require('../images/camera.png'),
            properties: {
                width: '100%'
            }
        });

        var cameraModifier = new StateModifier({
            origin: [0.5, 0],
            align: [0.5, 0],
            transform: Transform.behind
        });

        this.add(cameraModifier).add(camera);
        camera.on('deploy', function(){
            _createSlideshow.call(this);
        }.bind(this));
    }

    function _createSlideshow() {
        var slideshowView = new SlideshowView({
            size: [this.options.slideWidth, this.options.slideHeight],
            data: this.options.data
        });

        var slideshowModifier = new StateModifier({
            origin: [0.5, 0],
            align: [0.5, 0],
            transform: Transform.translate(0, this.options.slidePosition, 0)
        });

        var slideshowContainer = new ContainerSurface({
            properties: {
                overflow: 'hidden'
            }
        });

        this.add(slideshowModifier).add(slideshowContainer);
        slideshowContainer.add(slideshowView);
        slideshowContainer.context.setPerspective(1000);
    }

    module.exports = AppView;
});


src/data/SlideData.js
define(function(require, exports, module) {
    var SlideData = {
        picasaUrl: 'https://picasaweb.google.com/data/feed/api/all',
        queryParams: '?kind=photo&q=puppy&max-results=5&imgmax=720&alt=json',
        defaultImage: 'https://lh4.googleusercontent.com/-HbYp2q1BZfQ/U3LXxmWoy7I/AAAAAAAAAJk/VqI5bGooDaA/s1178-no/1.jpg'
    };
 
    SlideData.getUrl = function() {
        return SlideData.picasaUrl + SlideData.queryParams;
    };
 
    SlideData.parse = function(data) {
        var urls = [];
        data = JSON.parse(data);
        var entries = data.feed.entry;
        for (var i = 0; i < entries.length; i++) {
            var media = entries[i].media$group;
            urls.push(media.media$content[0].url);
        }
        return urls;
    };
 
    module.exports = SlideData;
});

src/views/SlideView.js
/*** SlideView.js ***/

define(function(require, exports, module) {
    var View = require('famous/core/View');
    var Surface = require('famous/core/Surface');
    var Transform = require('famous/core/Transform');
    var StateModifier = require('famous/modifiers/StateModifier');
    var ImageSurface = require('famous/surfaces/ImageSurface');
    var Transitionable   = require('famous/transitions/Transitionable');
    var SpringTransition = require('famous/transitions/SpringTransition');

    Transitionable.registerMethod('spring', SpringTransition);

    var SlideData = require('../data/SlideData');

    function SlideView() {
        View.apply(this, arguments);

        this.rootModifier = new StateModifier({
            size: this.options.size
        });

        this.mainNode = this.add(this.rootModifier);

        _createBackground.call(this);
        _createFilm.call(this);
        _createPhoto.call(this);
    }

    SlideView.prototype = Object.create(View.prototype);
    SlideView.prototype.constructor = SlideView;

    SlideView.DEFAULT_OPTIONS = {
        size: [400, 450],
        filmBorder: 15,
        photoBorder: 3,
        photoUrl: SlideData.defaultImage,
        angle: -0.5
    };

    function _createBackground() {
        var background = new Surface({
            properties: {
                backgroundColor: '#FFFFF5',
                boxShadow: '0 10px 20px -5px rgba(0, 0, 0, 0.5)',
                cursor: 'pointer'
            }
        });

        this.mainNode.add(background);

        background.on('click', function() {
            this._eventOutput.emit('click');
        }.bind(this));
    }

    function _createFilm() {
        this.options.filmSize = this.options.size[0] - 2 * this.options.filmBorder;

        var film = new Surface({
            size: [this.options.filmSize, this.options.filmSize],
            properties: {
                backgroundColor: '#222',
                zIndex: 1,
                pointerEvents: 'none'
            }
        });

        var filmModifier = new StateModifier({
            origin: [0.5, 0],
            align: [0.5, 0],
            transform: Transform.translate(0, this.options.filmBorder, 0.05)
        });

        this.mainNode.add(filmModifier).add(film);
    }

    function _createPhoto() {
        var size = this.options.filmSize - 2 * this.options.photoBorder;

        var photo = new ImageSurface({
            size: [size, size],
            content: this.options.photoUrl,
            properties: {
                zIndex: 2,
                pointerEvents: 'none'
            }
        });

        this.photoModifier = new StateModifier({
            origin: [0.5, 0],
            align: [0.5, 0],
            transform: Transform.translate(0, this.options.filmBorder + this.options.photoBorder, 0.1),
            opacity: 0.01
        });

        this.mainNode.add(this.photoModifier).add(photo);
    }

    SlideView.prototype.fadeIn = function() {
        this.photoModifier.setOpacity(1, { duration: 1500, curve: 'easeIn' });
        this.shake();
    };

    SlideView.prototype.shake = function() {
        this.rootModifier.halt();

        this.rootModifier.setTransform(
            Transform.rotateX(this.options.angle),
            { duration: 200, curve: 'easeOut' }
        );

        this.rootModifier.setTransform(
            Transform.identity,
            { method: 'spring', period: 600, dampingRatio: 0.15 }
        );
    };

    module.exports = SlideView;
});


src/views/SlideshowView.js
define(function(require, exports, module) {
    var View = require('famous/core/View');
    var Surface = require('famous/core/Surface');
    var Transform = require('famous/core/Transform');
    var StateModifier = require('famous/modifiers/StateModifier');
    var Lightbox = require('famous/views/Lightbox');
    var Easing = require('famous/transitions/Easing');

    var SlideView = require('./SlideView');

    function SlideshowView() {
        View.apply(this, arguments);

        this.rootModifier = new StateModifier({
            size: this.options.size,
            origin: [0.5, 0],
            align: [0.5, 0]
        });

        this.mainNode = this.add(this.rootModifier);

        _createLightbox.call(this);
        _createSlides.call(this);
    }

    SlideshowView.prototype = Object.create(View.prototype);
    SlideshowView.prototype.constructor = SlideshowView;

    SlideshowView.DEFAULT_OPTIONS = {
        size: [450, 500],
        data: undefined,
        lightboxOpts: {
            inOpacity: 1,
            outOpacity: 0,
            inOrigin: [0, 0],
            outOrigin: [0, 0],
            showOrigin: [0, 0],
            inTransform: Transform.thenMove(Transform.rotateX(0.9), [0, -300, -300]),
            outTransform: Transform.thenMove(Transform.rotateZ(0.7), [0, window.innerHeight, -1000]),
            inTransition: { duration: 650, curve: 'easeOut' },
            outTransition: { duration: 500, curve: Easing.inCubic }
        }
    };

    function _createLightbox() {
        this.lightbox = new Lightbox(this.options.lightboxOpts);
        this.mainNode.add(this.lightbox);
    }

    function _createSlides() {
        this.slides = [];
        this.currentIndex = 0;

        for (var i = 0; i < this.options.data.length; i++) {
            var slide = new SlideView({
                size: this.options.size,
                photoUrl: this.options.data[i]
            });

            this.slides.push(slide);

            slide.on('click', this.showNextSlide.bind(this));
        }

        this.showCurrentSlide();
    }

    SlideshowView.prototype.showCurrentSlide = function() {
        this.ready = false;

        var slide = this.slides[this.currentIndex];
        this.lightbox.show(slide, function() {
            this.ready = true;
            slide.fadeIn();
        }.bind(this));
    };

    SlideshowView.prototype.showNextSlide = function() {
        if (!this.ready) return;

        this.currentIndex++;
        if (this.currentIndex === this.slides.length) this.currentIndex = 0;
        this.showCurrentSlide();
    };

    module.exports = SlideshowView;
});


src/main.js
define(function(require, exports, module) {
    // load non-js resources (css, index.html)
    require('./resources');

    // import dependencies
    var Engine = require('famous/core/Engine');
    var Modifier = require('famous/core/Modifier');
    var Transform = require('famous/core/Transform');
    var ImageSurface = require('famous/surfaces/ImageSurface');
    var Utility = require('famous/utilities/Utility');

    var AppView = require('./views/AppView');
    var SlideData = require('./data/SlideData');

    // create the main context
    var mainContext = Engine.createContext();
    mainContext.setPerspective(1000);

    var contextSize;
    Engine.nextTick(function(){
        contextSize = mainContext.getSize();
        Utility.loadURL(SlideData.getUrl(), initApp);
    });

    function initApp(data) {
        data = SlideData.parse(data);
        var appWidth = (contextSize[0] > contextSize[1]) ? contextSize[1] : contextSize[0];
        var appView = new AppView({ data : data, cameraWidth: 0.5 * appWidth });

        mainContext.add(appView);
    }

    /**
     * Support for multiple configurations using the global ENV variable.
     *
     * Set the variable in webpack.config.js, or with the --env=XXXX flag.
     */
    if(ENV === "production") {
        console.log("Hooray! This is the production environment. You can use different settings!");
    } else {
        console.log("You're running in the "+ENV+" environment. Try \"webpack --env=production\".");
    }

    /**
     * Support for Cordova
     */
//    document.addEventListener("deviceready", function(){
//        // Add the cordova.js script tag to your HTML to load cordova.
//        //
//        //  <script type="text/javascript" src="cordova.js"></script>
//        console.log('Cordova loaded succesfully - Device Ready!');
//    }, false);
});
